src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java

```diff
@@ -226,18 +221,45 @@ private void setUserInfoInThreadContext(User user) {
         }
     }
 
-    public PrivilegesEvaluatorResponse evaluate(
-        final User user,
-        String action0,
-        final ActionRequest request,
-        Task task,
-        final Set<String> injectedRoles
-    ) {
+    public PrivilegesEvaluationContext createContext(User user, String action0, ActionRequest request, Task task, Set<String> injectedRoles)
+        throws PrivilegesEvaluatorResponse.NotAllowedException {
+        if (!isInitialized()) {
+            throw new OpenSearchSecurityException("OpenSearch Security is not initialized.");
+        }
+
+        TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
+        ImmutableSet<String> mappedRoles = ImmutableSet.copyOf((injectedRoles == null) ? mapRoles(user, caller) : injectedRoles);
+        final String injectedRolesValidationString = threadContext.getTransient(
+            ConfigConstants.OPENDISTRO_SECURITY_INJECTED_ROLES_VALIDATION
+        );
+        if (injectedRolesValidationString != null) {
+            HashSet<String> injectedRolesValidationSet = new HashSet<>(Arrays.asList(injectedRolesValidationString.split(",")));
+            if (!mappedRoles.containsAll(injectedRolesValidationSet)) {
+                PrivilegesEvaluatorResponse presponse = new PrivilegesEvaluatorResponse();
+                presponse.allowed = false;
+                presponse.missingSecurityRoles.addAll(injectedRolesValidationSet);
+                presponse.reason("Injected roles validation failed");
+                log.info("Roles {} are not mapped to the user {}", injectedRolesValidationSet, user);
+                throw new PrivilegesEvaluatorResponse.NotAllowedException(presponse);
+            }
+            mappedRoles = ImmutableSet.copyOf(injectedRolesValidationSet);
+        }
+
+        return new PrivilegesEvaluationContext(user, mappedRoles, action0, request, task, irr);
```

Comment:
why pass irr instead of irr.resolveRequest (which is what we utilize at the end)?

Reply:
This allows lazy execution. `irr.resolveRequest()` can be expensive, thus lazy execution helps us to skip it in cases where it is not needed.

Reply:
I'm still confused. Are we trying to avoid another call to irr.resolveRequest if the result if already non-null?

Reply:
Exactly. The call to `irr.resolveRequest()` can be amazingly expensive on clusters with many indices, so you want to avoid calling it without a good reason. That's why we have lazy execution to make sure that it is only executed when really needed. Additionally, if it is executed, we cache the reference in the `PrivilegeEvaluationContext` object to avoid having to call  `irr.resolveRequest()`  later again the resolved indices should be needed again.